src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -3,311 +3,820 @@
+
+            internal ITaskItem[] ReadItemGroup()
+            {
+                var items = new ITaskItem[_reader.ReadInt32()];
+
+                for (int i = 0; i < items.Length; i++)
+                {
+                    items[i] = ReadItem();
+                }
+
+                return items;
+            }
+
+            private ITaskItem ReadItem()
+            {
+                var item = new TaskItem(_reader.ReadString());
+                int metadataCount = _reader.ReadInt32();
+
+                for (int i = 0; i < metadataCount; i++)
+                {
+                    string key = _metadataStringTable[_reader.ReadInt32()];
+                    string value = _metadataStringTable[_reader.ReadInt32()];
+                    item.SetMetadata(key, value);
+                }
+
+                return item;
             }
         }
 
-        private ITaskItem[] RaisePackageAssets<T>(
-            LockFileTarget target, 
-            Func<LockFileTargetLibrary, IList<T>> getAssets,
-            Func<T, bool> filter = null,
-            Action<T, ITaskItem> setup = null) 
-            where T : LockFileItem
+        private sealed class CacheWriter : IDisposable
         {
-            var items = new List<ITaskItem>();
+            private const int InitialStringTableCapacity = 32;
+
+            private ResolvePackageAssets _task;
+            private BinaryWriter _writer;
+            private LockFile _lockFile;
+            private NuGetPackageResolver _packageResolver;
+            private LockFileTarget _compileTimeTarget;
+            private LockFileTarget _runtimeTarget;
+            private Dictionary<string, int> _stringTable;
+            private List<string> _metadataStrings;
+            private List<int> _bufferedMetadata;
+            private Placeholder _metadataStringTablePosition;
+            private int _itemCount;
+
+            public CacheWriter(ResolvePackageAssets task)
+            {
+                var targetFramework = NuGetUtils.ParseFrameworkName(task.TargetFrameworkMoniker);
+
+                _task = task;
+                _lockFile = new LockFileCache(task.BuildEngine4).GetLockFile(task.ProjectAssetsFile);
+                _packageResolver = NuGetPackageResolver.CreateResolver(_lockFile, _task.ProjectPath);
+                _compileTimeTarget = _lockFile.GetTargetAndThrowIfNotFound(targetFramework, runtime: null);
+                _runtimeTarget = _lockFile.GetTargetAndThrowIfNotFound(targetFramework, _task.RuntimeIdentifier);
+                _stringTable = new Dictionary<string, int>(InitialStringTableCapacity, StringComparer.Ordinal);
+                _metadataStrings = new List<string>(InitialStringTableCapacity);
+                _bufferedMetadata = new List<int>();
+
+                Directory.CreateDirectory(Path.GetDirectoryName(task.ProjectAssetsCacheFile));
+                var stream = File.Open(task.ProjectAssetsCacheFile, FileMode.Create, FileAccess.ReadWrite, FileShare.None);
```

Comment:
Do you need to worry about concurrent writes here? I see you have None there, but if it happens, will this bubble up? Is that what you intend?

Reply:
The cache file is in Intermediate directory, and has msbuildprojectname in it. If we double write here, project configuration is busted. We'd double write a bunch of other stuff too.

Reply:
Sounds good.